### PR TITLE
Update controllers to achieve mini-kubemove functionalities

### PR DESCRIPTION
**Pull request by hossainemruz**
**Tasks:**
- [x] Update MoveEngine controller
- [x] Update DataSync controller

### CODE REVIEW

File: GNUmakefile [hunk 1]
```diff
@@ -103,11 +103,17 @@ dummy_plugin-image: dummy_plugin
...
-
 base-image:
 	@echo "Building base docker image for kubemove"
 	@docker build -t $(BASE_ENGINE_IMG):$(BASE_ENGINE_TAG) -f build/Dockerfile-base ./build
 
+push-images:
+	@docker push $(ENGINE_IMG):$(IMG_TAG)
+	@docker push $(PAIR_IMG):$(IMG_TAG)
+	@docker push $(DS_IMG):$(IMG_TAG)
+	@docker push $(PLUGIN_IMG):$(IMG_TAG)
```
Comment (mynktl) [inline GNUmakefile]: `deploy-images` pushes the image. `push-images` not required. And We don't need to upload plugin image. `dummy_plugin` is for reference only, to write a plugin.

File: GNUmakefile [hunk 2]
```diff
@@ -244,3 +250,466 @@ deploy-images:
...
+				sleep 20 																						   && \
+				mc config host add my-minio http://$(MINIO_SERVER_ADDRESS) $(MINIO_ACCESS_KEY) $(MINIO_SECRET_KEY) && \
+				mc mb my-minio/$(BUCKET_NAME)  																		  \
+			'
+	@echo "Done"
+
+# Install necessary Kubemove resources
+# Example: make prepare REGISTRY=<your docker registry>
+.PHONY: prepare
+prepare: register-crds create-rbac-resources deploy-mp-ctrl deploy-me-ctrl deploy-ds-ctrl deploy-minio create-minio-bucket
```
Comment (mynktl) [inline GNUmakefile]: Can we limit `prepare` to installation of kubemove only. So We will have
`deploy_kubemove` : install kubemove crd, namespace, rbac, controllers
`purge_kubemove` : remove crd, namespace, rbac, controllers
MinIO related stuff we can remove from makefile and have it under ./examples/minio.

File: GNUmakefile [hunk 3]
```diff
@@ -244,3 +250,466 @@ deploy-images:
 	@IMAGE=$(ENGINE_IMG) ./build/push
 	@IMAGE=$(PAIR_IMG) ./build/push
 	@IMAGE=$(DS_IMG) ./build/push
+	@echo ""
+	@echo "Pushing developer image...."
+	@docker push $(DEV_IMAGE)
```
Comment (mynktl) [inline GNUmakefile]: We don't need to push `DEV_IMAGE`, right? It is used during build only. So we can generate the DEV_IMAGE first and proceed with the build.

File: GNUmakefile [hunk 4]
```diff
@@ -244,3 +250,466 @@ deploy-images:
...
+			-v $$(pwd)/.go/cache:/.cache                                 \
+			-v $(HOME)/.kube:/.kube                                      \
+			-w /go/src/$(MODULE)                                         \
+			--env HTTP_PROXY=$(HTTP_PROXY)                               \
+			--env HTTPS_PROXY=$(HTTPS_PROXY)                             \
+			--env GO111MODULE=on                                         \
+			--env GOFLAGS="-mod=vendor"                                  \
+			--env KUBECONFIG=$(INTERNAL_CONFIG)                          \
+			$(DEV_IMAGE)                                                 \
+			/bin/bash -c '                                               \
```
Comment (mynktl) [inline GNUmakefile]: This will not change anything locally, right? We can run this command without docker also.

Comment (mynktl) [reply GNUmakefile]: For other kubemove installation command also.

File: pkg/controller/moveengine/controller.go [hunk 5]
```diff
@@ -126,48 +134,211 @@ func (r *ReconcileMoveEngine) Reconcile(request reconcile.Request) (reconcile.Re
...
+		if requeue {
+			return reconcile.Result{Requeue: requeue, RequeueAfter: requeueAfter}, nil
+		}
+	case v1alpha1.MoveEngineInitializationFailed:
+		// MoveEngine has failed to be initialized. There is no point of processing further.
+		r.log.Info("Skipping processing MoveEngine: %s/%s. Reason: MoveEngine has failed to be initialized.", instance.Namespace, instance.Name)
+		return reconcile.Result{}, nil
+	case v1alpha1.MoveEngineReady:
+		// MoveEngine is ready. Now, we can start rest of the sync process.
+
```
Comment (mynktl) [inline pkg/controller/moveengine/controller.go]: It will be better if we move below code to a separate function.

**Review by mynktl — CHANGES_REQUESTED** [review]
Looks good. I've given few comments about Makefile. Let's remove minio completely from Makefile and move it to `examples/resources/repository/minio/`. We can have `examples/resources/repository/minio/README.md` about installation steps for minio.

File: cmd/kubemove/main.go [hunk 6]
```diff
@@ -9,11 +9,11 @@ func main() {
...
 
 	/*
-	// TODO
-	root := plugin.NewCmdKubeMove()
-	if err := root.Execute(); err != nil {
-		os.Exit(1)
-	}
-	 */
+		// TODO
+		root := plugin.NewCmdKubeMove()
```
Comment (mynktl) [inline cmd/kubemove/main.go]: Let's remove `cmd/kubemove` and also from `.travis.yml`.

File: GNUmakefile [hunk 7]
```diff
@@ -244,3 +250,466 @@ deploy-images:
 	@IMAGE=$(ENGINE_IMG) ./build/push
 	@IMAGE=$(PAIR_IMG) ./build/push
 	@IMAGE=$(DS_IMG) ./build/push
+	@echo ""
+	@echo "Pushing developer image...."
```
Comment (mynktl) [inline GNUmakefile]: Let's add following targets to `.travis.yml`:
- make dev-image
- make lint
- make format

Comment (mynktl) [reply GNUmakefile]: Can you merge existing `golint` to `lint`
https://github.com/kubemove/kubemove/blob/4a3cf00edc9853dcf40e23836a8b47506bee945d/GNUmakefile#L50

File: GNUmakefile (hunk 1 above)
Comment (mynktl) [reply GNUmakefile]: `./push` script is taking care of git tag. So we can remove
https://github.com/kubemove/kubemove/pull/33/files#diff-b0fa74059a31d3efac99fea6d20ca25aL29-L39

Comment (hossainemruz) [reply GNUmakefile]: removed

File: GNUmakefile (hunk 2 above)
Comment (hossainemruz) [reply GNUmakefile]: Done. Also, Minio resources has been moved in `elasticsearch-plugin` repo.

File: GNUmakefile (hunk 3 above)
Comment (hossainemruz) [reply GNUmakefile]: Agreed. Removed.

File: GNUmakefile (hunk 4 above)
Comment (hossainemruz) [reply GNUmakefile]: removed

File: pkg/controller/moveengine/controller.go (hunk 5 above)
Comment (hossainemruz) [reply pkg/controller/moveengine/controller.go]: done

File: cmd/kubemove/main.go (hunk 6 above)
Comment (hossainemruz) [reply cmd/kubemove/main.go]: removed

File: GNUmakefile (hunk 7 above)
Comment (hossainemruz) [reply GNUmakefile]: done

**Review by mynktl — DISMISSED** [review]

**Comment by mynktl** [comment]
@hossainemruz  In travis build there are some errors about `kubectl`. https://github.com/kubemove/kubemove/pull/33/files/c8ba2a5bb5d57321dffdaf9beea65a86dac846ff#diff-b0fa74059a31d3efac99fea6d20ca25aR239-R242 Changes are required to create MovePair only. So better to move it under `create_local_mp`

**Comment by hossainemruz** [comment]
> @hossainemruz In travis build there are some errors about `kubectl`. [kubemove/kubemove/pull/33/files/c8ba2a5bb5d57321dffdaf9beea65a86dac846ff#diff-b0fa74059a31d3efac99fea6d20ca25aR239-R242](https://github.com/kubemove/kubemove/pull/33/files/c8ba2a5bb5d57321dffdaf9beea65a86dac846ff#diff-b0fa74059a31d3efac99fea6d20ca25aR239-R242) Changes are required to create MovePair only. So better to move it under `create_local_mp`

Done.

**Review by mynktl — APPROVED** [review]

**Pull request merged**